File: script/make/protocol/lua/FriendProtocol.lua
FriendProtocol = {}

function FriendProtocol.encode(offset, protocol, data)
    if protocol == 11501 then
        local table = {}

        return table
    elseif protocol == 11502 then
        local table = {}
        -- 好友角色ID
        table[offset] = string.pack(">I8", data)
        offset = offset + 1
        return table
    elseif protocol == 11503 then
        local table = {}
        -- 好友角色ID
        table[offset] = string.pack(">I8", data)
        offset = offset + 1
        return table
    elseif protocol == 11504 then
        local table = {}
        -- 好友角色ID
        table[offset] = string.pack(">I8", data)
        offset = offset + 1
        return table
    elseif protocol == 11505 then
        local table = {}
        -- 好友角色ID
        table[offset] = string.pack(">I8", data)
        offset = offset + 1
        return table
    elseif protocol == 11506 then
        local table = {}
        -- 好友角色ID
        table[offset] = string.pack(">I8", data)
        offset = offset + 1
        return table
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

function FriendProtocol.decode(offset, protocol, bytes)
    if protocol == 11501 then
        -- 
        local data = {}
        local dataLength = string.unpack(">I2", bytes, offset)
        offset = offset + 2
        for dataIndex = 1, dataLength do
            -- 
            -- 好友角色ID
            local dataDataFriendRoleId = string.unpack(">I8", bytes, offset)
            offset = offset + 8
            -- 好友名字
            local dataDataFriendName = string.unpack(">s2", bytes, offset)
            offset = offset + 2 + string.len(dataDataFriendName)
            -- 关系状态(申请:1/好友:2/黑名单:3)
            local dataDataRelation = string.unpack(">I1", bytes, offset)
            offset = offset + 1
            -- 添加/修改状态时间
            local dataDataTime = string.unpack(">I4", bytes, offset)
            offset = offset + 4
            -- object
            local dataData = {friendRoleId = dataDataFriendRoleId, friendName = dataDataFriendName, relation = dataDataRelation, time = dataDataTime}
            data[dataIndex] = dataData
        end
        return data
    elseif protocol == 11502 then
        -- 结果
        local data = string.unpack(">s2", bytes, offset)
        offset = offset + 2 + string.len(data)
        return data
    elseif protocol == 11503 then
        -- 结果
        local data = string.unpack(">s2", bytes, offset)
        offset = offset + 2 + string.len(data)
        return data
    elseif protocol == 11504 then
        -- 
        -- 结果
        local dataResult = string.unpack(">s2", bytes, offset)
        offset = offset + 2 + string.len(dataResult)
        -- 好友角色ID
        local dataFriendRoleId = string.unpack(">I8", bytes, offset)
        offset = offset + 8
        -- object
        local data = {result = dataResult, friendRoleId = dataFriendRoleId}
        return data
    elseif protocol == 11505 then
        -- 
        -- 结果
        local dataResult = string.unpack(">s2", bytes, offset)
        offset = offset + 2 + string.len(dataResult)
        -- 好友角色ID
        local dataFriendRoleId = string.unpack(">I8", bytes, offset)
        offset = offset + 8
        -- object
        local data = {result = dataResult, friendRoleId = dataFriendRoleId}
        return data
    elseif protocol == 11506 then
        -- 
        -- 结果
        local dataResult = string.unpack(">s2", bytes, offset)
        offset = offset + 2 + string.len(dataResult)
        -- 好友角色ID
        local dataFriendRoleId = string.unpack(">I8", bytes, offset)
        offset = offset + 8
        -- object
        local data = {result = dataResult, friendRoleId = dataFriendRoleId}
        return data
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

File: script/make/protocol/cs/FriendProtocol.cs
public static class FriendProtocol
{
    public static void Encode(System.Text.Encoding encoding, System.IO.BinaryWriter writer, System.UInt16 protocol, System.Object data) 
    {
        switch (protocol) 
        {
            case 11501:
            {
                return;
            }
            case 11502:
            {
                // 好友角色ID
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int64)(System.UInt64)data));
                return;
            }
            case 11503:
            {
                // 好友角色ID
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int64)(System.UInt64)data));
                return;
            }
            case 11504:
            {
                // 好友角色ID
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int64)(System.UInt64)data));
                return;
            }
            case 11505:
            {
                // 好友角色ID
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int64)(System.UInt64)data));
                return;
            }
            case 11506:
            {
                // 好友角色ID
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int64)(System.UInt64)data));
                return;
            }
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }

    public static System.Object Decode(System.Text.Encoding encoding, System.IO.BinaryReader reader, System.UInt16 protocol) 
    {
        switch (protocol) 
        {
            case 11501:
            {
                // 
                var dataLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var data = new System.Collections.Generic.List<System.Object>(dataLength);
                while (dataLength-- > 0)
                {
                    // 
                    // 好友角色ID
                    var friendRoleId = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                    // 好友名字
                    var friendNameLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    var friendName = encoding.GetString(reader.ReadBytes(friendNameLength));
                    // 关系状态(申请:1/好友:2/黑名单:3)
                    var relation = reader.ReadByte();
                    // 添加/修改状态时间
                    var time = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                    // object
                    var friend = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"friend_role_id", friendRoleId}, {"friend_name", friendName}, {"relation", relation}, {"time", time}};
                    // add
                    data.Add(friend);
                }
                return data;
            }
            case 11502:
            {
                // 结果
                var dataLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var data = encoding.GetString(reader.ReadBytes(dataLength));
                return data;
            }
            case 11503:
            {
                // 结果
                var dataLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var data = encoding.GetString(reader.ReadBytes(dataLength));
                return data;
            }
            case 11504:
            {
                // 
                // 结果
                var resultLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var result = encoding.GetString(reader.ReadBytes(resultLength));
                // 好友角色ID
                var friendRoleId = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                // object
                var data = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"result", result}, {"friend_role_id", friendRoleId}};
                return data;
            }
            case 11505:
            {
                // 
                // 结果
                var resultLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var result = encoding.GetString(reader.ReadBytes(resultLength));
                // 好友角色ID
                var friendRoleId = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                // object
                var data = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"result", result}, {"friend_role_id", friendRoleId}};
                return data;
            }
            case 11506:
            {
                // 
                // 结果
                var resultLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var result = encoding.GetString(reader.ReadBytes(resultLength));
                // 好友角色ID
                var friendRoleId = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                // object
                var data = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"result", result}, {"friend_role_id", friendRoleId}};
                return data;
            }
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }
}